From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GKI Fake BL <gki@fakebl>
Date: Fri, 9 May 2026 00:00:00 +0000
Subject: [PATCH] fs/proc: Fake bootloader lock state via /proc/cmdline

Modify /proc/cmdline at kernel level to fake:
- androidboot.flash.locked=1
- androidboot.vbmeta.device_state=locked
- androidboot.verifiedbootstate=green

This ensures init sets the correct ro.boot.* properties without needing resetprop.
---
 fs/proc/cmdline.c | 38 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 37 insertions(+), 1 deletion(-)

diff --git a/fs/proc/cmdline.c b/fs/proc/cmdline.c
index e851568..a8c1a88 100644
--- a/fs/proc/cmdline.c
+++ b/fs/proc/cmdline.c
@@ -4,10 +4,38 @@
 #include <linux/kernel.h>
 #include <linux/proc_fs.h>
 #include <linux/seq_file.h>
+#include <linux/slab.h>
+#include <linux/string.h>
 
 static int cmdline_proc_show(struct seq_file *m, void *v)
 {
-	seq_puts(m, saved_command_line);
+	char *cmdline = kstrdup(saved_command_line, GFP_KERNEL);
+	char *ptr;
+
+	if (cmdline) {
+		ptr = strstr(cmdline, "androidboot.flash.locked=0");
+		if (ptr)
+			ptr[sizeof("androidboot.flash.locked=") - 1] = '1';
+
+		ptr = strstr(cmdline, "androidboot.vbmeta.device_state=unlocked");
+		if (ptr) {
+			size_t off = sizeof("androidboot.vbmeta.device_state=") - 1;
+			memmove(ptr + off + 6, ptr + off + 8, strlen(ptr + off + 8) + 1);
+			memcpy(ptr + off, "locked", 6);
+		}
+
+		ptr = strstr(cmdline, "androidboot.verifiedbootstate=orange");
+		if (ptr) {
+			size_t off = sizeof("androidboot.verifiedbootstate=") - 1;
+			memmove(ptr + off + 5, ptr + off + 6, strlen(ptr + off + 6) + 1);
+			memcpy(ptr + off, "green", 5);
+		}
+
+		seq_puts(m, cmdline);
+		kfree(cmdline);
+	} else {
+		seq_puts(m, saved_command_line);
+	}
 	seq_putc(m, '\n');
 	return 0;
 }
-- 
2.43.0

